jdbc-driver/src/main/java/com/google/cloud/dataproc/jdbc/DataprocInfo.java

```diff
@@ -0,0 +1,169 @@
+import java.util.*;
+import java.util.stream.Collectors;
+
+/** Helper class to get cluster info through Dataproc API. */
+public class DataprocInfo {
+
+    private static final String HIVE_PROTOCOL = "jdbc:hive2://";
+    private final ClusterControllerClient clusterControllerClient;
+    private final HiveJdbcConnectionOptions params;
+
+    // Constructor dependency injection
+    public DataprocInfo(HiveJdbcConnectionOptions params, ClusterControllerClient controller) {
+        this.params = params;
+        this.clusterControllerClient = controller;
+    }
+
+    /**
+     * Format client passed in url that is accepted by Dataproc to be Hive acceptable format.
+     * jdbc:hive2://<host>:<port>/<dbName>;transportMode=http;httpPath=<http_endpoint>;<otherSessionConfs>?<hiveConfs>#<hiveVars>
+     *
+     * @return
+     * @throws InvalidURLException
+     */
+    public String toHiveJdbcUrl() throws InvalidURLException {
+        String hiveJdbcURL =
+                String.format(
+                        "%s%s:%d/%s;transportMode=%s;httpPath=%s",
+                        HIVE_PROTOCOL,
+                        getHost(),
+                        params.port(),
+                        params.dbName(),
+                        params.transportMode(),
+                        params.httpPath());
+        if (params.otherSessionConfs() != null) {
+            hiveJdbcURL = String.format("%s;%s", hiveJdbcURL, params.otherSessionConfs());
+        }
+        if (params.hiveConfs() != null) {
+            hiveJdbcURL = String.format("%s?%s", hiveJdbcURL, params.hiveConfs());
+        }
+        if (params.hiveVars() != null) {
+            hiveJdbcURL = String.format("%s#%s", hiveJdbcURL, params.hiveVars());
+        }
+        return hiveJdbcURL;
+    }
+
+    /**
+     * Use Dataproc client libraries to get cluster.
+     *
+     * @return name of the cluster's master node
+     * @throws InvalidURLException
+     */
+    public String getHost() throws InvalidURLException {
+        String host;
+        if (params.clusterName() != null) {
+            host = getClusterByName();
+        } else {
+            // clusterPoolLabel can be null, but filter is not empty string since
+            // it at least has "status.state = RUNNING"
+            String filter = formatClusterFilterString();
+            host = findClusterInPool(filter);
+        }
+        return host + "-m";
+    }
+
+    /**
+     * Supports finding & verifying the cluster indicated by name.
+     *
+     * @return the cluster of the given name
+     * @throws InvalidURLException
+     */
+    public String getClusterByName() throws InvalidURLException {
+        try {
+            Cluster cluster =
+                    clusterControllerClient.getCluster(
+                            params.projectId(), params.region(), params.clusterName());
+            return cluster.getClusterName();
+        } catch (ApiException e) {
+            throw new InvalidURLException(
```

Comment:
Seems like `InvalidURLException` is not an appropriate exception type to throw here? Is failure always caused by invalid URL?